sopel/irc/backends.py

```diff
@@ -389,16 +391,78 @@ async def _run_forever(self) -> None:
         # open connection
         try:
             self._reader, self._writer = await asyncio.open_connection(
-                **self.get_connection_kwargs(),
+                **connection_kwargs,
             )
-        except ssl.SSLError:
-            LOGGER.exception('Unable to connect due to SSL error.')
+
+        # SSL Error
+        except ssl.SSLCertVerificationError as err:
+            LOGGER.error(
+                'Unable to connect due to an SSL validation error: %s',
+                err,
+            )
+            self.log_exception()
+            # tell the bot to quit without restart
+            self.bot.hasquit = True
+            self.bot.wantsrestart = False
+            return
+        except ssl.SSLError as err:
+            LOGGER.error('Unable to connect due to an SSL error: %s', err)
+            self.log_exception()
+            # tell the bot to quit without restart
+            self.bot.hasquit = True
+            self.bot.wantsrestart = False
+            return
+
+        # Specific connection error
+        except socket.gaierror as err:
+            LOGGER.error(
+                'Unable to connect due to invalid IRC server address: %s',
+                err,
+            )
+            LOGGER.error(
+                'You should verify that "%s:%s" is the correct address '
+                'to connect to the IRC server.',
+                connection_kwargs.get('host'),
+                connection_kwargs.get('port'),
+            )
+            self.log_exception()
             # tell the bot to quit without restart
             self.bot.hasquit = True
             self.bot.wantsrestart = False
             return
-        except Exception:
-            LOGGER.exception('Unable to connect.')
+        except TimeoutError as err:
+            LOGGER.error('Unable to connect due to a timeout: %s', err)
+            self.log_exception()
+            # tell the bot to quit without restart
+            self.bot.hasquit = True
+            self.bot.wantsrestart = False
+            return
+
+        # Generic connection error (OSError is used for any connection error)
+        except OSError as err:
+            LOGGER.error(
+                'Unable to connect: %s',
+                err,
+            )
+            LOGGER.error(
+                'You should verify that "%s:%s" is the correct address '
+                'to connect to the IRC server.',
```

Comment:
When I tested various error case, yes, some OSError are caused by wrong host:port. For example, I got different OSError based on:

* does the domain name exist?
* is the domain name valid (i.e. doesn't contain invalid characters)?
* is the port valid?
* is the port available?
* is there something to accept the connection on the other side?

That's why I had OSError separated from other errors: because some errors have specific exception class, and some have just OSError.

As I mentioned, the next improvement would be to list all the OSError's error code and log a different message for them.

Reply:
To give more example, I tested with a typo in a valid domain name (adding a letter here and there, forgetting a dot somewhere), and I tested with the wrong port number, and I got different exceptions with different error code, and both can be resolved by checking and fixing the host:port in the configuration file.

Reply:
Sounds like you surveyed the potential issues pretty well and it's good advice. Consider it resolved! :grin:

Reply:
Counterpoint: Subclasses of [`ConnectionError`](https://docs.python.org/3.7/library/exceptions.html#ConnectionError) seem unlikely to be solved by checking host:port. Or are we guessing that the most likely reason for something like `ConnectionRefusedError` is that the user e.g. mistyped a port number?

At any rate, since `TimeoutError` is allowed to have its own `except` block, the other obviously-connection-related subclasses of `OSError` could also get their own. (And if checking host:port _is_ relevant for `ConnectionError` and friends, it's also relevant for `TimeoutError`.)

Reply:
It's not relevant to `TimeoutError` because it usually means something is wrong in the network, either on the client side or the server side. It's an error because a system call ended in a timeout. And `TimeoutError` isn't a subclass of `ConnectionError`:

> Subclasses are [BrokenPipeError](https://docs.python.org/3/library/exceptions.html#BrokenPipeError), [ConnectionAbortedError](https://docs.python.org/3/library/exceptions.html#ConnectionAbortedError), [ConnectionRefusedError](https://docs.python.org/3/library/exceptions.html#ConnectionRefusedError) and [ConnectionResetError](https://docs.python.org/3/library/exceptions.html#ConnectionResetError).

From [ConnectionError](https://docs.python.org/3/library/exceptions.html#ConnectionError).

However, none of these would need the help of `host:port`, so I will add a generic `except ConnectionError`. That'll left the `OSError` and some can be solved with the correct `host:port` (which my manual tests show).

That should be better with this change!